File: sdk/ruby/lib/rocksky/resources/mirror.rb
module Rocksky
  module Resources
    # `app.rocksky.mirror.*` endpoints. Require an authenticated client.
    class Mirror < Base
      # List configured mirror sources for the authenticated user.
      def get_mirror_sources
        query("app.rocksky.mirror.getMirrorSources")
      end

      # Configure a mirror source (e.g. Last.fm, ListenBrainz).
      def put_mirror_source(provider:, enabled: nil, external_username: nil, api_key: nil)
        body = {
          provider: provider,
          enabled: enabled,
          externalUsername: external_username,
          apiKey: api_key
        }.compact
        procedure("app.rocksky.mirror.putMirrorSource", body: body)
      end
    end
  end
end

File: sdk/python/src/rocksky/resources/mirror.py
"""``app.rocksky.mirror.*`` — external mirroring (Last.fm, ListenBrainz, …)."""

from __future__ import annotations

from typing import Any

from ..models import MirrorSource
from ._base import Resource, parse_list


class MirrorResource(Resource):
    async def list_sources(self) -> list[MirrorSource]:
        data = await self._transport.query(
            "app.rocksky.mirror.getMirrorSources", auth=True
        )
        if isinstance(data, list):
            return [MirrorSource.model_validate(item) for item in data]
        return parse_list(MirrorSource, data, key="sources") or parse_list(
            MirrorSource, data, key="mirrorSources"
        )

    async def put_source(
        self,
        provider: str,
        *,
        enabled: bool | None = None,
        external_username: str | None = None,
        api_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"provider": provider}
        if enabled is not None:
            body["enabled"] = enabled
        if external_username is not None:
            body["externalUsername"] = external_username
        if api_key is not None:
            body["apiKey"] = api_key
        return await self._transport.procedure(
            "app.rocksky.mirror.putMirrorSource", body=body, auth=True
        ) or {}
